File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example2/StatusFilterIterator.cs
namespace IteratorDesignPattern.Example2;

public class StatusFilterIterator : IIterator<Order>
{
    private readonly OrderCollection _collection;
    private readonly OrderStatus _filterStatus;
    private int _currentIndex = 0;

    public StatusFilterIterator(OrderCollection collection, OrderStatus filterStatus)
    {
        _collection = collection;
        _filterStatus = filterStatus;
    }

    public bool HasNext()
    {
        while (_currentIndex < _collection.Count)
        {
            if (_collection[_currentIndex].Status == _filterStatus)
            {
                return true;
            }
            _currentIndex++;
        }
        return false;
    }

    public Order Next()
    {
        if (!HasNext())
        {
            throw new InvalidOperationException("No more orders matching the status filter.");
        }

        return _collection[_currentIndex++];
    }

    public void Reset()
    {
        _currentIndex = 0;
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example2/src/main/java/behavioral/iterator/example2/StatusFilterIterator.java
package behavioral.iterator.example2;

public class StatusFilterIterator implements IIterator<Order> {
    private final OrderCollection collection;
    private final OrderStatus filterStatus;
    private int currentIndex = 0;

    public StatusFilterIterator(OrderCollection collection, OrderStatus filterStatus) {
        this.collection = collection;
        this.filterStatus = filterStatus;
    }

    @Override
    public boolean hasNext() {
        while (currentIndex < collection.getCount()) {
            if (collection.get(currentIndex).getStatus() == filterStatus) return true;
            currentIndex++;
        }
        return false;
    }

    @Override
    public Order next() {
        if (!hasNext()) throw new java.util.NoSuchElementException();
        return collection.get(currentIndex++);
    }

    @Override
    public void reset() { currentIndex = 0; }
}
